### DISCUSSION TITLE
Organizing complex states for many different views

### DISCUSSION THREAD

**Discussion by randall-coding** [hecrj/iced#836] [Q&A] [answered]
I'm making an application with many different views.  Currently all my states are stored in my Application root object in a struct like so:

```
pub struct States{ 
    pub user_button_states: Vec <button::State>,
    pub back_button: button::State,
    
    pub channel_menu: Vec <button::State>,
    pub scroll: scrollable::State,
    
    pub text_input: text_input::State,
    pub input_text: String,
    
    pub filter_radio_buttons: String,
    pub table_order: Option<String>,
    pub header_buttons: Vec<button::State>,
}
```
Some of these states are shared across views, others are not.  But I'm about to change this to something more like this:

```
pub struct States{
  pub shared: SharedStates, 
  pub page1: Page1States,
  pub page2: Page2States,
  pub page3: Page3States,
  pub page4: Page4States,
  pub page5: Page5States,
}
```  

I think it would be really convenient to break this up by view page, each page getting it's own set of states and it's own struct.  But now I'm wondering as my project grows, is it good for performance to instantiate every state for every page when the application loads?  And is there another way to do this?

Let's say I have 20 pages, then I will need to store a struct with 20 other structs within it.  Or if there is 100 pages.  Might this cause some issues?

**Comment by hecrj** [comment, marked as answer]
You probably do not need the state of all the different pages ready at once. I recommend you to use an enum to model the state of your pages. For instance:

```rust
mod home;
mod login;
mod account;

pub use home::Home;
pub use login::Login;
pub use account::Account;

pub enum Page {
    Home(Home),
    Login(Login),
    Account(Account),
    // ...
}
```

**Comment by randall-coding** [comment]
Thanks.  So I would use the enum by generating my page states for a given page after I received a message to switch pages in my update function, then assign this new generated state to my page enum?  If I understand the process correctly.  That is what makes sense to me right now anyway.
